### PR TITLE
update gate time series method to allow selection of gating type

### PR DESCRIPTION
**Pull request by ahnitz**
This should make it easier to select gate method in higher level codes in the future (i.e. pycbc inspiral / pycbc live).  This PR also addes a dummy to time/frequency series so that functions can be made generic of accepting a time or frequency series where possible. 

Depends on #3371 

'paint' method
download (4)

'taper' method
download (5)

'hard' method
download (6)

### CODE REVIEW

**Comment by ahnitz** [comment]
Plots made by this script
```
import pylab
from pycbc.catalog import Merger
m = Merger("GW170817")
ts = m.strain('L1')

for method in ['hard', 'taper', 'paint']:
    ts2 = ts.gate(m.time-1.0, window=0.15, method=method)      
    w = ts2.whiten(2, 2).time_slice(m.time-5, m.time+4)
    t, f, p = w.qtransform(.001, logfsteps=100, 
                           frange=(30, 1000), qrange=(100, 100))

    pylab.figure(figsize=[15, 6])
    pylab.pcolormesh(t-m.time, f, p, vmax=20)
    pylab.colorbar()
    pylab.yscale('log')
```

**Comment by ahnitz** [comment]
@spxiwh @cdcapano @titodalcanton As expected, the 'paint' method does a better job of not ringing up loud lines (if the PSD is long enough to resolve them), while not requiring the tapering portion of the gate.

**Comment by spxiwh** [comment]
Hmm ... Looking at the plot for the "paint" method, I'm thinking that it doesn't really seem to work as the power still drops to 0 in the gap. Wasn't the idea to fully inpaint with noise so that this should just look like Gaussian noise?

**Comment by ahnitz** [comment]
@spxiwh No, that's not the idea. Might be easier to explain in the chat.

**Comment by titodalcanton** [comment]
Thanks, this is cool. Some minor comments:
* Can we add a comment with a reference to the IAS method description to the code that implements the in-painting?
* What is the motivation for having the "hard" method? I would not expect it to be useful in practice. Maybe something in pycbc_inference?
* Is the "hard" method basically the "taper" method with a tapering window set to zero? If so, can the "hard" case in `TimeSeries.gate()` just call `gate_data()` with a window set to zero, or even be made a sub-case of the "taper" method?

**Comment by ahnitz** [comment]
> Thanks, this is cool. Some minor comments:
> 
> * Can we add a comment with a reference to the IAS method description to the code that implements the in-painting?

Ok, will do. 

> * What is the motivation for having the "hard" method? I would not expect it to be useful in practice. Maybe something in pycbc_inference?

It makes sense to simulate true gaps in the data. I don't imagine it would be really used in the same circumstances where one would naturally use the others. It does however serve as a natural 'worst case' in this comparison. 

> * Is the "hard" method basically the "taper" method with a tapering window set to zero? If so, can the "hard" case in `TimeSeries.gate()` just call `gate_data()` with a window set to zero, or even be made a sub-case of the "taper" method?

That is right, though, given it is only a couple lines to invoke, I don't think this would be any clearer.

**Comment by ahnitz** [comment]
@cdcapano @titodalcanton I've just update the PR, please re-review

File: pycbc/types/timeseries.py [hunk 1]
```diff
@@ -478,25 +478,56 @@ def psd(self, segment_duration, **kwds):
...
+            from pycbc.strain.gate import gate_and_paint
+            if invpsd is None:
+                # These are some bare minimum settings, normally you
+                # should probably provide a psd
+                invpsd = 1. / self.filter_psd(self.duration/32, self.delta_f, 0)
+            lindex = int((time - window - self.start_time) / self.delta_t)
+            rindex = lindex + int(2 * window / self.delta_t)
+            lindex = lindex if lindex >=0 else 0
+            rindex = rindex if rindex <= len(self) else len(self)
+            return gate_and_paint(data, lindex, rindex, invpsd, copy=False)
```
Comment (cdcapano) [inline pycbc/types/timeseries.py:524]: This should be `copy=not inplace`.

Comment (ahnitz) [reply pycbc/types/timeseries.py:524]: Done

Comment (cdcapano) [reply pycbc/types/timeseries.py:524]: Huh? `copy` is still hard-coded to `False` here.

Comment (cdcapano) [reply pycbc/types/timeseries.py:524]: My original comment was referring to the fact that the copy keyword argument in the call to `gate_and_paint` wasn't respecting the `inplace` argument, not that you needed to change the name of `inplace` to `copy`.

Comment (ahnitz) [reply pycbc/types/timeseries.py:524]: Ok, then I misunderstood, and so did you. This is correct. The copy is done at the top for all the types. @cdcapano  Look at line 508

Comment (cdcapano) [reply pycbc/types/timeseries.py:524]: Oh! Sorry, I missed that. Looks fine then.

**Pull request merged**